controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -186,29 +208,37 @@ private void triggerReadyJobs(LockedApplication application) {
      */
     private boolean changesAvailable(Application application, JobStatus previous, JobStatus next) {
         if ( ! application.deploying().isPresent()) return false;
-        Change change = application.deploying().get();
 
-        if ( ! previous.lastSuccess().isPresent()) return false;
+        Change change = application.deploying().get();
 
-        if (change instanceof Change.VersionChange) {
+        if (change instanceof Change.VersionChange) { // Propagate upgrade while making sure we never downgrade
             Version targetVersion = ((Change.VersionChange)change).version();
-            if ( ! (targetVersion.equals(previous.lastSuccess().get().version())) )
-                return false; // version is outdated
-            // The below is checked again in allowedTriggering, right before actual triggering.
-            if (next != null && isOnNewerVersionInProductionThan(targetVersion, application, next.type()))
-                return false; // Don't downgrade
-        }
 
-        if (next == null) return true;
-        if ( ! next.lastSuccess().isPresent()) return true;
+            // Is the target version tested?
+            if ( ! lastSuccessfulIs(targetVersion, JobType.stagingTest, application))
+                return false;
+
+            // Is it this job's turn to upgrade?
+            if ( previous.type().isProduction() && ! isOnAtLeastProductionVersion(targetVersion, application, previous.type()))
+                return false;
+
+            // Is the next a test zone which already completed this upgrade successfully?
+            if (next != null && next.type().isTest() && lastSuccessfulIs(targetVersion, next.type(), application))
+                return false;
+
+            // Is the next a production job for a zone already upgraded to this version or a newer one?
+            if (next != null && next.type().isProduction() && isOnAtLeastProductionVersion(targetVersion, application, next.type()))
+                return false;
 
-        JobStatus.JobRun previousSuccess = previous.lastSuccess().get();
-        JobStatus.JobRun nextSuccess = next.lastSuccess().get();
-        if (previousSuccess.revision().isPresent() && ! previousSuccess.revision().equals(nextSuccess.revision()))
-            return true;
-        if ( ! previousSuccess.version().equals(nextSuccess.version()))
             return true;
-        return false;
+        }
+        else { // revision changes do not need to handle downgrading
+            if ( ! previous.lastSuccess().isPresent()) return false;
+            if (next == null) return true;
+            if ( ! next.lastSuccess().isPresent()) return true;
+            return previous.lastSuccess().get().revision().isPresent() &&
+                   ! previous.lastSuccess().get().revision().equals(next.lastSuccess().get().revision());
```

Comment:
We really should separate out verification from deployment, as comparing apples one place and oranges another, while making sure the conditions remain the same, is rather straining ...